[CALCITE-1253]Create Elasticsearch adapter for Calcite

sbcd90 opened the pull request:
Kindly review & accept this contribution.

sbcd90 committed:
[CALCITE-1253]Create Elasticsearch adapter for Calcite

michaelmior commented:
For all the other databases currently included in the VM, we've used Puppet for provisioning. Unless there's a really good reason not to, I think it would be nice to stick with this. @vlsi thoughts? There's a [well-supported module for ElasticSearch](https://github.com/elastic/puppet-elasticsearch) that should make this pretty painless.

vlsi commented:
@michaelmior , frankly speaking I'm fine with both approaches provided they can be maintained.

Puppet approach worked fine so far.

Current PR looks fine for me modulo some details like `apt-get purge openjdk*` (==remove all jdk, thus it might impact other DBs),  "missing end of line at the end of the files", "java 7 being used" (why not java 8?)

vlsi commented on vm/install_elasticsearch.sh:
```diff
@@ -0,0 +1,27 @@
+sudo apt-get update
+
+# install OpenJDK7
+sudo apt-get purge openjdk*
```
This looks dangerous. Is it really required to uninstall all JDK versions?
It looks like it could (==will) impact other JVM-based DBs.

vlsi commented on vm/install_elasticsearch.sh:
```diff
@@ -0,0 +1,27 @@
+
+# install OpenJDK7
+sudo apt-get purge openjdk*
+sudo apt-get -y install openjdk-7-jdk
```
Is java 7 mandatory for ElasticSearch? I would rather prefer having java 8.

vlsi commented on vm/init_elasticsearch.sh:
```diff
@@ -0,0 +1,20 @@
+# This file is called automatically by Vagrant VM provision
+echo Importing zips dataset
+cd /dataset
+curl -s -XPOST localhost:9200/_bulk --data-binary "@zips/zips_es.json"
```
Having newline at the end of the file would be nice.

michaelmior commented:
@sbcd90 @vlsi It was fairly quick to whip up something which uses Puppet for provisioning instead. I think this should address the outstanding issues. Just doing some additional testing and I'll push the code shortly.

michaelmior commented:
My branch with ES provisioning using Puppet is [here](https://github.com/michaelmior/calcite-test-dataset/tree/esadapter).

michaelmior commented:
Personally I think it would be nice to stick with Puppet for consistency. I've been doing some more testing and everything seems to work fine. However, I think the memory allocation of the VM needs to be bumped again. The Druid integration tests failed until I changed the VM memory allocation from 3072 to 4096.

sbcd90 commented:
Hello @vlsi @michaelmior , Thanks for the review.

@vlsi , I have made all the changes mentioned by you in the review.
- `apt-get purge openjdk*` is removed.
- updated to openjdk8
- added a newline.

@michaelmior , would request @vlsi & you to decide which approarch is better & close this issue. The current changes in this PR do not fail Druid integration tests. 
Then only the other dependent PR( https://github.com/apache/calcite/pull/236 ) can be reviewed.
Thanks in advance.

michaelmior commented:
@sbcd90 There's nothing you did that broke the Druid tests but available memory has recently started to become an issue on the VM. As new services get added, it will occasionally be necessary to bump up the memory allocation.

@vlsi I'll leave you to decide what to merge although as I mentioned earlier, I think it would be nice to stick with Chef. My branch should be ready for a squash merge as-is although we may want to consider increasing the memory limit. Could always wait to see if the CI builds fail.

vlsi commented:
> I think it would be nice to stick with Chef

@michaelmior , did you want to say "stick with Puppet"?
Let's stick with puppet.

https://github.com/michaelmior/calcite-test-dataset/tree/esadapter looks good

michaelmior commented:
Sorry, yes I meant stick with Puppet :) @sbcd90 Can you try my branch and makes sure it works as expected? Your integration tests pass, so I think it should be fine.

sbcd90 commented:
@michaelmior I tried out your branch & it works fine for me as well.

michaelmior commented:
Opened a new PR with my commit on top. Left the memory limit as-is for now and we can bump later if needed.
